pkg/plugins/golang/v4/scaffolds/internal/templates/main.go

```diff
@@ -121,12 +127,54 @@ const (
 			os.Exit(1)
 		}
 	}
+`
+
+	checkAPIKindExistsCodeFragment = `checkAPIKindExists("%s")
+`
+
+	// nolint:lll
+	checkAPIKindExistsMethodCodeFragment = `// checkAPIKindExists checks if a specific kind is available on the cluster across all groups and versions.
```

Comment:
I would not make use of this method.

Exiting is not good practice I would argue, as this will likely just cause the Pod to be recreated and then crashed again (ChrashLoop).

I think the reconciler will log errors automatically when it should watch a non-existing resource. Having these logs should be sufficient and feels like the way to go here:
https://github.com/kubernetes-sigs/controller-runtime/blob/659afe993be9b2b5c67be7d3e1d20397a91061ce/pkg/internal/source/kind.go#L68-L72

Reply:
Using the Kubernetes Discovery API in an Operator for external APIs ensures the external CRD is installed before the controller starts, preventing runtime errors. It dynamically checks for the required resources, allowing the operator to fail gracefully if CRDs are missing.

This approach improves the operator's robustness by ensuring external dependencies are met, which aligns with Kubebuilder’s goal of helping developers build reliable, high-quality solutions while reducing the learning curve. Some references on using third-party APIs can be found [here](https://developers.redhat.com/blog/2020/02/04/how-to-use-third-party-apis-in-operator-sdk-projects#step_4__use_the_api_in_the_controllers).

Therefore, I would strongly advocate for keeping this implementation. Otherwise, admins might have a very hard time debugging the solution or understanding why it’s not functioning correctly when it should fail during deployment due to unmet prerequisites. Either we lost a great opportunity to help users know how they can do things.

Reply:
Hi @fischor 

IHMO we should add the validation.
**However, it need to be improved we should check the GKV and not only the Kind** 
Therefore, I am not so happy with how it is done yet
So, we might can either let this part for an follow up where we discuss only the new markers and the inject code.